File: Haskell/p002-Basic-Neuron-Layer.hs
multWeights :: [Double] -> [Double] -> Double  -> Double
multWeights i w b = sum (zipWith (*) i w) + b

layer :: [Double] -> [[Double]] -> [Double] -> [Double]
layer i ws bs = zipWith (multWeights i) ws bs 

inputs :: [Double]
inputs = [1.0, 2.0, 3.0, 2.5]

weights1 :: [Double]
weights1 = [0.2, 0.8, -0.5, 1.0]
weights2 :: [Double]
weights2 = [0.5, -0.91, 0.26, -0.5]
weights3 :: [Double]
weights3 = [-0.26, -0.27, 0.17, 0.87]

weights :: [[Double]]
weights = [weights1, weights2, weights3]

bias1 :: Double
bias1 = 2.0
bias2 :: Double
bias2 = 3.0
bias3 :: Double
bias3 = 0.5

biases :: [Double]
biases = [bias1, bias2, bias3]

output = layer inputs weights biases

main = print output

File: Perl/p002-Basic-Neuron-Layer.pl
@inputs = (1.0, 2.0, 3.0, 2.5);

@weights1 = (0.2, 0.8, -0.5, 1.0);
@weights2 = (0.5, -0.91, 0.26, -0.5);
@weights3 = (-0.26, -0.27, 0.17, 0.87);

$bias1 = 2.0;
$bias2 = 3.0;
$bias3 = 0.5;

@output = ($inputs[0]*$weights1[0] + $inputs[1]*$weights1[1] + $inputs[2]*$weights1[2] + $inputs[3]*$weights1[3] + $bias1,
          $inputs[0]*$weights2[0] + $inputs[1]*$weights2[1] + $inputs[2]*$weights2[2] + $inputs[3]*$weights2[3] + $bias2,
          $inputs[0]*$weights3[0] + $inputs[1]*$weights3[1] + $inputs[2]*$weights3[2] + $inputs[3]*$weights3[3] + $bias3);

print "[$output[0], $output[1], $output[2]]\n";
